primitives/src/merkle_tree/field_based_mht/optimized/mod.rs

```diff
@@ -0,0 +1,568 @@
+use algebra::Field;
+use crate::{BatchFieldBasedMerkleTreeParameters, BatchFieldBasedHash, FieldBasedMerkleTree, FieldBasedMerkleTreePath, FieldBasedMHTPath, FieldBasedHash, FieldBasedHashParameters, check_precomputed_parameters};
+use std::marker::PhantomData;
+
+/// An implementation of FieldBasedMerkleTree, optimized in time and memory,
+/// and able to support any BatchFieldBasedHash and Merkle arity.
+/// TODO: Test with arity > 2
+#[derive(Clone)]
+pub struct FieldBasedOptimizedMHT<T: BatchFieldBasedMerkleTreeParameters>{
+    root: T::Data,
+    // Stores all MT nodes
+    array_nodes: Vec<T::Data>,
+    processing_step: usize,
+    // Stores the initial index of each level of the MT
+    initial_pos: Vec<usize>,
+    // Stores the final index of each level of the MT
+    final_pos: Vec<usize>,
+    // Stores the index up until the nodes were already hashed, for each level
+    processed_pos: Vec<usize>,
+    // Stores the next available index for each level of the MT
+    new_elem_pos: Vec<usize>,
+    rate: usize,
+    height: usize,
+    finalized: bool,
+    _tree_parameters: PhantomData<T>,
+}
+
+impl<T: BatchFieldBasedMerkleTreeParameters> FieldBasedOptimizedMHT<T> {
+
+    /// Creates a new tree given its `height` and `processing_step`, used to tune
+    /// the memory usage of the tree. In particular, `processing_step` defines the
+    /// number of leaves to store before triggering the computation of the root.
+    /// Decreasing `processing_step` leads to less memory consumption but
+    /// significantly worsen performances, as the computation of the root is
+    /// triggered more often, along with a batch hash that may perform poorly when
+    /// there are few elements to batch; of course, conversely, increasing
+    /// `processing_step` increases the memory usage too but improves performances.
+    /// Meaningful values for `processing_step` are between 1 (i.e. update the root
+    /// at each leaf), leading to best memory efficiency but worse performances, and
+    /// the maximum number of leaves (or the mean number of leaves you plan to add),
+    /// leading to worse memory efficiency but best performances (root is computed
+    /// just once, but all the leaves must be kept in memory).
```

Comment:
I think the root is always computed once but when the processing_step is large, most of the hashes is done with batched hashes which is faster.